src/main/validateParameters.js

```diff
@@ -0,0 +1,123 @@
+  const stackLines = err.stack.split('\n');
+  
+  // The third line in the stack trace should contain the caller function name
+  // The format is usually like "    at functionName (file:line:column)"
+  const callerLine = stackLines[3];
+  console.log(JSON.stringify(callerLine))
+  // Extract the function name from the stack trace line
+  const match = callerLine.trim().match(/^at (\w+)\.(\w+)/);
+  
+  // Return the function name if it exists
+  return match ? JSON.stringify(match[0]) : undefined;
+}
+
+/**
+ * Validates that the keys and types of the properties in `realObj` match the specifications in `expectedObj`.
+ *
+ * @param {Object} realObj - The object to be validated.
+ * @param {Object} expectedObj - An object where each key corresponds to a key in `realObj`, 
+ * and the value is either a string representing a primitive type (e.g., "string", "number", "boolean") 
+ * or a constructor function (class) representing the expected type.
+ * @param {boolean} [prohibitNull=true] - Whether to prohibit null values.
+ * @param {boolean} [checkCallingUserData=true] - Whether to check for calling user data.
+ *
+ * @throws {EMSError} If a key in `expectedObj` is missing from `realObj`.
+ * @throws {EMSError} If the type of a property in `realObj` does not match the expected type in `expectedObj`.
+ * @throws {EMSError} If an unsupported type specification is encountered in `expectedObj`.
+ *
+ *
+ * @example
+ * // Primitive types validation
+ * const realObj = { name: "John", age: 30, isAdmin: true };
+ * const expectedObj = { name: "string", age: "number", isAdmin: "boolean" };
+ * validateParameters(realObj, expectedObj); // pass validation
+ *
+ * @example
+ * // Class instance validation
+ * class Person {
+ *   constructor(name, age) {
+ *     this.name = name;
+ *     this.age = age;
+ *   }
+ * }
+ * const realObj = { profile: new Person("John", 30) };
+ * const expectedObj = { profile: Person };
+ * validateParameters(realObj, expectedObj); // pass validation
+ */
+function validateParameters(realObj, expectedObj, prohibitNull = true, checkCallingUserData = true) {
+
+  const callingFunctionName = getCallingFunctionName()
+  if (!realObj) throw new EMSError(ERROR_MSGS.NULL_OBJECT(callingFunctionName), ERROR_CODES.NULL_OBJECT);
+
+  // Add callingUserData to expectedObj if needed
+  if (checkCallingUserData) {
+    expectedObj = {
+      ...expectedObj,
+      callingUser: {
+        username: PRIMITIVE_TYPES.STRING,
+        type: PRIMITIVE_TYPES.STRING,
+      },
+    };
+  }
+
+  for (const key in expectedObj) {
+    if (expectedObj.hasOwnProperty(key)) {
+      // Check if the key exists in realObj
+      if (!realObj.hasOwnProperty(key)) {
+        throw new EMSError(ERROR_MSGS.MISSING_KEY(key,callingFunctionName), ERROR_CODES.MISSING_KEY);
+      }
+
+      // Get the expected type
+      const expectedType = expectedObj[key];
+      const realValue = realObj[key];
+
+      if (prohibitNull && realValue === null) {
```

Comment:
What if some input is mandatory and some optional?
would I need to call this function twice?

Reply:
Actually yes.
I hope we wont get to this at all, it will be a total mess, and super hard to maintain if we allow only some fields to be null and the other not.

I made it so we can choose (with null or without null, for all fields) and change it easily in the future, but it will be really bad if we will do a mix